File: core/src/main/java/dev/morphia/mapping/conventions/kotlin/KotlinPropertyDiscovery.java
package dev.morphia.mapping.conventions.kotlin;

import dev.morphia.Datastore;
import dev.morphia.mapping.MapperOptions.Builder;
import dev.morphia.mapping.MappingException;
import dev.morphia.mapping.codec.pojo.EntityModelBuilder;
import dev.morphia.mapping.conventions.MorphiaConvention;
import org.bson.codecs.pojo.PropertyAccessor;

import java.lang.reflect.Field;
import java.util.Arrays;

/**
 * @morphia.internal
 * @since 2.2
 */
public class KotlinPropertyDiscovery implements MorphiaConvention {
    private final Builder optionsBuilder;

    public KotlinPropertyDiscovery(Builder optionsBuilder) {
        this.optionsBuilder = optionsBuilder;
    }

    @Override
    public void apply(Datastore datastore, EntityModelBuilder builder) {
        Field field = Arrays.stream(builder.getType().getDeclaredFields())
                            .filter(f -> f.getName().equals("$$delegatedProperties"))
                            .findFirst().get();

        if (field != null) {
            field.setAccessible(true);
            builder.propertyModelByName(field.getName() + "$delegate")
                   .name(field.getName())
                   .discoverMappedName(optionsBuilder.build())
                   .accessor(new ReadWritePropertyAccessor(field));
        }
    }
}

class ReadWritePropertyAccessor implements PropertyAccessor<Object> {
    private final Field field;

    ReadWritePropertyAccessor(Field kProperty) {
        this.field = kProperty;
        field.setAccessible(true);
    }

    @Override
    public <S> Object get(S instance) {
        try {
            return field.get(instance);
        } catch (ReflectiveOperationException e) {
            throw new MappingException(e.getMessage(), e);
        }
    }

    @Override
    public <S> void set(S instance, Object value) {
        try {
            field.set(instance, value);
        } catch (ReflectiveOperationException e) {
            throw new MappingException(e.getMessage(), e);
        }
    }
}

File: core/src/main/kotlin/dev/morphia/mapping/conventions/kotlin/KotlinPropertyDiscovery.kt
package dev.morphia.mapping.conventions.kotlin

import dev.morphia.Datastore
import dev.morphia.mapping.MapperOptions.Builder
import dev.morphia.mapping.codec.pojo.EntityModelBuilder
import dev.morphia.mapping.conventions.MorphiaConvention
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import org.bson.codecs.pojo.PropertyAccessor
import kotlin.reflect.KMutableProperty
import kotlin.reflect.KProperty
import kotlin.reflect.jvm.javaField
import kotlin.reflect.jvm.javaGetter
import kotlin.reflect.jvm.javaSetter

class KotlinPropertyDiscovery(private val optionsBuilder: Builder) : MorphiaConvention {
    @Suppress("UNCHECKED_CAST")
    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE")
    override fun apply(datastore: Datastore, builder: EntityModelBuilder) {
        val field = builder.type.declaredFields.firstOrNull { it.name == "\$\$delegatedProperties" }
        if (field != null) {
            field.trySetAccessible()
            for (kProperty in field.get(builder.type) as Array<KProperty<*>>) {
                builder.propertyModelByName("${kProperty.name}\$delegate")
                    .name(kProperty.name)
                    .discoverMappedName(optionsBuilder.build())
                    .accessor(ReadWritePropertyAccessor(kProperty as KMutableProperty<*>))
            }
        }
    }
}

class ReadWritePropertyAccessor(private val kProperty: KMutableProperty<*>) : PropertyAccessor<Any> {
    init {
        kProperty.javaField?.isAccessible = true
        kProperty.javaGetter?.isAccessible = true
        kProperty.javaSetter?.isAccessible = true
    }

    override fun <S : Any?> get(instance: S): Any? {
        val getter = kProperty.getter
        val value = getter.call(instance)
        return value
    }

    override fun <S : Any?> set(instance: S, value: Any?) {
        kProperty.setter.call(instance, value)
    }
}
